subsys/net/lib/dns/mdns_responder.c

```diff
@@ -0,0 +1,520 @@
+	offset = net_pkt_get_len(pkt) - data_len;
+
+	/* Store the DNS query name into a temporary net_buf. This means
+	 * that largest name we can resolve is CONFIG_NET_BUF_DATA_SIZE
+	 * which typically is 128 bytes. This is done using net_buf so that
+	 * we do not increase the stack usage of RX thread.
+	 */
+	result = net_pkt_get_data(ctx, BUF_ALLOC_TIMEOUT);
+	if (!result) {
+		ret = -ENOMEM;
+		goto quit;
+	}
+
+	/* TODO: Instead of this temporary copy, just use the net_pkt directly.
+	 */
+	ret = net_frag_linear_copy(dns_data, pkt->frags, offset, data_len);
+	if (ret < 0) {
+		goto quit;
+	}
+
+	dns_msg.msg = dns_data->data;
+	dns_msg.msg_size = data_len;
+
+	if (dns_header_qr(dns_msg.msg) != DNS_QUERY) {
+		ret = -EINVAL;
+		goto quit;
+	}
+
+	if (dns_header_opcode(dns_msg.msg) != DNS_QUERY) {
+		ret = -EINVAL;
+		goto quit;
+	}
+
+	ret = dns_unpack_query_header(&dns_msg, NULL);
+	if (ret < 0) {
+		ret = -EINVAL;
+		goto quit;
+	}
+
+	queries = ret;
+
+	NET_DBG("Received %d %s from %s", queries,
+		queries > 1 ? "queries" : "query",
+		net_pkt_family(pkt) == AF_INET ?
+		net_sprint_ipv4_addr(&NET_IPV4_HDR(pkt)->src) :
+		net_sprint_ipv6_addr(&NET_IPV6_HDR(pkt)->src));
+
+	do {
+		enum dns_rr_type qtype;
+		enum dns_class qclass;
+		u8_t *lquery;
+
+		memset(result->data, 0, net_buf_tailroom(result));
+		result->len = 0;
+
+		ret = dns_unpack_query(&dns_msg, result, &qtype, &qclass);
+		if (ret < 0) {
+			goto quit;
+		}
+
+		/* Handle only .local queries */
+		lquery = strrchr(result->data + 1, '.');
+		if (!lquery || memcmp(lquery, (const void *){ ".local" }, 7)) {
+			continue;
+		}
+
+		NET_DBG("[%d] query %s/%s label %s (%d bytes)", queries,
+			qtype == DNS_RR_TYPE_A ? "A" : "AAAA", "IN",
+			result->data, ret);
+
+		/* If the query matches to our hostname, then send reply.
+		 * We skip the first dot, and make sure there is dot after
+		 * matching hostname.
+		 */
+		if (!strncasecmp(hostname, result->data + 1, hostname_len) &&
+		    (result->len - 1) >= hostname_len &&
+		    &(result->data + 1)[hostname_len] == lquery) {
+			NET_DBG("mDNS query to our hostname %s.local",
+				hostname);
+			send_response(ctx, pkt, result, qtype);
+		}
+	} while (--queries);
+
+	if (result) {
+		net_pkt_frag_unref(result);
+	}
+
+	return 0;
+
+quit:
+	if (result) {
+		net_pkt_frag_unref(result);
+	}
+
+	return ret;
+}
+
+static void recv_cb(struct net_context *net_ctx,
+		    struct net_pkt *pkt,
+		    int status,
+		    void *user_data)
+{
+	struct net_context *ctx = user_data;
+	struct net_buf *dns_data = NULL;
+	struct dns_addrinfo info = { 0 };
+	int ret;
+
+	ARG_UNUSED(net_ctx);
+	NET_ASSERT(ctx == net_ctx);
+
+	if (!pkt) {
+		return;
+	}
+
+	if (status) {
+		goto quit;
+	}
+
+	dns_data = net_buf_alloc(&dns_msg_pool, BUF_ALLOC_TIMEOUT);
+	if (!dns_data) {
+		goto quit;
+	}
+
+	ret = dns_read(ctx, pkt, dns_data, &info);
+	if (ret < 0 && ret != -EINVAL) {
+		NET_DBG("mDNS read failed (%d)", ret);
+		goto quit;
+	}
+
+quit:
```

Comment:
looks like quit label should be put at line 404, all the relevant gotos come from places where dns_data is not set.